### ISSUE TITLE
True matrices

### ISSUE THREAD

**Issue by niomaster** [utwente-fmt/vercors#791]
We need real multi-dimensional array support, since C requires it, and it would be nice to have in PVL. Let's also deal with the linearized matrix indexing problem at the same time (i.e. `ar[i*N+j]` where `ar.length == N*M, i < M, j < N`)

**Comment by dzagieboylo** [comment]
I've been looking into this as it is needed by lots of CUDA programs (e.g., tiling via statically sized `__shared__` memory is very common).

There were some bugs (e.g., the type conversion of `CTArray` always flattened multi-dimensional arrays and didn't accurately track the inner pointer type) that I've fixed in my fork; however, I'm still running into a *major issue*.

The declaration `int arr[4]` gets translated to:
```
arr = new pointer<int>[4];
```
Which lets us establish facts about the (one-D) array's locations and permissions.

However, the declaration `int arr[4][2]` also just gets translated to a one liner:

```
arr = new pointer<pointer<int>>[4];
```
Which doesn't capture the fact that the inner arrays are also automatically allocated and write-permissible in the C semantics when provided static bounds.

I've added a translation which automatically generates this initialization code in `LangCToCol`  but it's actually ALSO difficult to generate the information for VerCors that all of the pointers are unique and correctly allocated; this makes using MD arrays impractical.

------

Is it possible to have better backend support for MD arrays that automatically establishes the write permissions for each location? I'm also happy to try to implement it, if given some pointers where to look.

**Comment by superaxander** [comment]
Yes we need support proper support for (multi-dimensional) arrays still. Treating them as pointers is incorrect since for example the following program verifies:

```c
int main(void) {
    int a = 10;
    int b[2][3];

    b[0] = &a;

    return 0;
}
```

but it doesn't compile since b[0] is of type int[] which cannot be assigned a pointer (in fact it is not assignable at all)

Another example:

```
struct A {
    int a[10];
};

int main(void) {
    struct A b;
    int c[10];
    b.a = c;

    return 0;
}
```

Yeah arrays are not pointers and we should not treat them as such. It's one of the things on my to do list. 

To have some rudimentary support for multi-dimensional arrays it would be fine for now to do something like the translation you mentioned and then a user could use something like `\matrix` but for pointers which would automatically include all the necessary quantifiers to show that all the elements are unique.

**Comment by superaxander** [comment]
I'm having a closer look at a proper encoding for arrays right now.
